### PR TITLE
Feature/wallet export import

### PR DESCRIPTION
**Pull request by keichiri**
The first version of wallet export/import. 

I will do some improvements meanwhile (more careful reading from Read while import, its behavior when blocking, and some other stuff). Also, I will write more tests for various scenarios

Please review it carefully and see if it fits your ideas.

Export/Import header is in the following format:
<wallet_header_length><version><timestamp (unix)><encryption_scheme_length><encryption_scheme><nonce_length><nonce><sha256 of everything before>

Putting encryption scheme in the header might be an overkill for now, but if we decide to let user pick encryption method, this is a must

### CODE REVIEW

File: libindy/src/api/mod.rs [hunk 1]
```diff
@@ -116,6 +116,9 @@ pub enum ErrorCode
     // Returned if provided wallet query is invalid
     WalletQueryError = 214,
 
+    // Returned if wallet import has failed
+    WalletImportError = 215,
```
Comment (vimmerru) [inline libindy/src/api/mod.rs]: I don't like this error. It doesn't describe anything about the cause.

File: libindy/src/services/wallet/mod.rs [hunk 2]
```diff
@@ -516,36 +518,66 @@ impl WalletService {
...
+    }
+
     pub const PREFIX: &'static str = "Indy::";
 
     fn add_prefix(&self, type_: &str) -> String {
         format!("{}{}", WalletService::PREFIX, type_)
     }
 }
 
+fn length_to_bigend_bytes(length: usize) -> [u8; 4] {
```
[the remark this answers is not in the document]
Comment (keichiri) [reply libindy/src/services/wallet/mod.rs]: yes, I will change it to u32

[the remark this answers is not in the document]
Comment (keichiri) [reply libindy/src/services/wallet/mod.rs]: Actually, this was left from before. I just forgot to clean it up. I will remove it now

Not used anywhere. All conversions are in export_import.rs

fn u64_to_bytes(u: u64) -> [u8; 8] {
    let u = u.to_be();
    let res: [u8; 8] = unsafe { mem::transmute(u) };
    res
}

fn u32_to_bytes(u: u32) -> [u8; 4] {
    let u = u.to_be();
    let res: [u8; 4] = unsafe { mem::transmute(u) };
    res
}

fn u16_to_bytes(u: u16) -> [u8; 2] {
    let u = u.to_be();
    let res: [u8; 2] = unsafe { mem::transmute(u) };
    res
}

File: libindy/src/api/mod.rs [hunk 3]
```diff
@@ -116,6 +116,15 @@ pub enum ErrorCode
     // Returned if provided wallet query is invalid
     WalletQueryError = 214,
 
+    // Returned if I/O error occurred during wallet operation
+    WalletIOError = 215,
```
Comment (vimmerru) [inline libindy/src/api/mod.rs]: We should use CommonIOError here

File: libindy/src/api/mod.rs [hunk 4]
```diff
@@ -116,6 +116,15 @@ pub enum ErrorCode
...
+
+    // Returned if wallet input has invalid structure (currently used in wallet import)
+    WalletStructureError = 216,
```
Comment (vimmerru) [inline libindy/src/api/mod.rs]: We should use CommonInvalidStructure here

File: libindy/src/errors/wallet.rs [hunk 5]
```diff
@@ -37,6 +37,9 @@ pub enum WalletError {
     ItemNotFound,
     ItemAlreadyExists,
     QueryError(String),
+    IOError(io::Error),
```
Comment (vimmerru) [inline libindy/src/errors/wallet.rs]: There is CommonError enum in WalletError already

**Comment by vimmerru** [comment]
I don't see any new API added in this import/export PR. Also there are no any integration tests. We don't merge the code without integration tests (at least basic). Also there are some additional comments that require handling. Do you consider this work as complete or plan to add new functions to api.rs?

File: libindy/src/api/wallet.rs [hunk 6]
```diff
@@ -297,6 +297,53 @@ pub extern fn indy_list_wallets(command_handle: i32,
...
+/// #Returns
+/// Error code
+///
+/// #Errors
+/// Common*
+/// Wallet*
+#[no_mangle]
+pub extern fn indy_export_wallet(command_handle: i32,
+                                 handle: i32,
+                                 export_path: *const c_char,
```
Comment (vimmerru) [inline libindy/src/api/wallet.rs]: I suggest to replace 

> export_path: *const c_char,
> export_key: *const c_char,

with export_config_json to allow extend export configuration in a backward compatible way

File: libindy/include/indy_wallet.h [hunk 7]
```diff
@@ -216,6 +216,70 @@ extern "C" {
...
+    ///
+    /// #Returns
+    /// Error code
+    ///
+    /// #Errors
+    /// Common*
+    /// Wallet*
+
+    extern indy_error_t indy_export_wallet(indy_handle_t  command_handle,
+                                           indy_handle_t  handle,
```
Comment (vimmerru) [inline libindy/include/indy_wallet.h]: > handle 

it is better to rename to wallet_handle

File: libindy/include/indy_wallet.h [hunk 8]
```diff
@@ -216,6 +216,70 @@ extern "C" {
...
                                           );
 
+    /// Exports opened wallet's content using key and path provided in export_config_json
+    ///
+    /// #Params
+    /// handle: wallet handle returned by indy_open_wallet.
+    /// export_config_json: JSON containing settings for input operation.
+    ///   {
+    ///     "path": path of the file that contains exported wallet content
+    ///     "key": passphrase used to export key
```
Comment (vimmerru) [inline libindy/include/indy_wallet.h]: > passphrase used to export key

passphrase used to derive export key

File: libindy/include/indy_wallet.h [hunk 9]
```diff
@@ -216,6 +216,70 @@ extern "C" {
...
+    /// pool_name: Name of the pool that corresponds to this wallet.
+    /// name: Name of the wallet.
+    /// xtype(optional): Type of the wallet. Defaults to 'default'.
+    ///                  Custom types can be registered with indy_register_wallet_type call.
+    /// config(optional): Wallet configuration json. List of supported keys are defined by wallet type.
+    ///                    if NULL, then default config will be used.
+    /// credentials: Wallet credentials json
+    ///   {
+    ///       "key": string,
+    ///       "rekey": Optional<string>,
```
Comment (vimmerru) [inline libindy/include/indy_wallet.h]: >  "rekey": Optional<string>,

We don't need rekey here

File: libindy/include/indy_wallet.h (hunk 7 above)
Comment (keichiri) [reply libindy/include/indy_wallet.h]: i used existing conventions, same as in close_wallet

File: libindy/src/api/mod.rs [hunk 10]
```diff
@@ -105,7 +105,7 @@ pub enum ErrorCode
     WalletStorageError = 210,
 
     // Error during encryption-related operations
-    WalletEncryptonError = 211,
+    WalletEncryptionError = 211,
```
Comment (vimmerru) [inline libindy/src/api/mod.rs]: > WalletEncryptionError 

I suggest to remove it in favor of:

- WalletAccessError if key provided by use is invalid
- ComminInvalidStructure if wallet export file has structure problems

File: libindy/src/api/mod.rs [hunk 11]
```diff
@@ -116,6 +116,18 @@ pub enum ErrorCode
     // Returned if provided wallet query is invalid
     WalletQueryError = 214,
 
+    // Returned if wallet input has invalid structure (currently used in wallet import)
+    WalletStructureError = 215,
```
Comment (vimmerru) [inline libindy/src/api/mod.rs]: Please replace it with ComminInvalidStructure

File: libindy/src/api/mod.rs [hunk 12]
```diff
@@ -116,6 +116,18 @@ pub enum ErrorCode
...
+
+    // Returned if the wallet is not empty, but it is required (currently only when importing)
+    WalletNotEmpty = 216,
```
Comment (vimmerru) [inline libindy/src/api/mod.rs]: Seems we don't need it for now as always import to new wallet

File: libindy/src/api/mod.rs [hunk 13]
```diff
@@ -116,6 +116,18 @@ pub enum ErrorCode
...
+
+    // Returned if path provided to wallet export already exists
+    WalletExportPathExists = 217,
```
Comment (vimmerru) [inline libindy/src/api/mod.rs]: I suggest to just raise CommonIOError here

File: libindy/src/api/mod.rs [hunk 14]
```diff
@@ -116,6 +116,18 @@ pub enum ErrorCode
...
+
+    // Returned if path provided to wallet import does not exist
+    WalletImportPathDoesNotExist = 218,
```
Comment (vimmerru) [inline libindy/src/api/mod.rs]: I suggest to just raise CommonIOError here

File: libindy/src/api/wallet.rs [hunk 15]
```diff
@@ -308,6 +308,132 @@ pub extern fn indy_list_wallets(command_handle: i32,
...
+///
+/// #Returns
+/// Error code
+///
+/// #Errors
+/// Common*
+/// Wallet*
+#[no_mangle]
+pub extern fn indy_export_wallet(command_handle: i32,
+                                 handle: i32,
```
Comment (vimmerru) [inline libindy/src/api/wallet.rs]: > handle

Rename it to wallet_handle

File: libindy/src/api/wallet.rs [hunk 16]
```diff
@@ -308,6 +308,132 @@ pub extern fn indy_list_wallets(command_handle: i32,
...
+// storage_type(optional): Type of the wallet storage. Defaults to 'default'.
+//                  Custom storage types can be registered with indy_register_wallet_storage_call
+/// config(optional): Wallet configuration json.
+///   {
+///       "storage": <object>  List of supported keys are defined by wallet type.
+///   }
+/// credentials: Wallet credentials json (if NULL, then default config will be used).
+///   {
+///       "key": string,
+///       "rekey": Optional<string>,
```
Comment (vimmerru) [inline libindy/src/api/wallet.rs]: We don't need rekey here

File: libindy/src/api/mod.rs (hunk 10 above)
Comment (keichiri) [reply libindy/src/api/mod.rs]: Its an old error, used throughout wallet, not added now

File: libindy/src/api/mod.rs (hunk 12 above)
Comment (keichiri) [reply libindy/src/api/mod.rs]: Oh yes thats correct. Impossible to get now.

File: libindy/src/api/wallet.rs (hunk 15 above)
Comment (keichiri) [reply libindy/src/api/wallet.rs]: i used existing conventions from close_wallet call

File: libindy/src/api/mod.rs (hunk 13 above)
Comment (keichiri) [reply libindy/src/api/mod.rs]: I wouldnt say its IO error. It is more of an configuration error

File: libindy/src/api/mod.rs (hunk 14 above)
Comment (keichiri) [reply libindy/src/api/mod.rs]: I wouldnt say its IO error. It is more of an configuration error

File: libindy/src/services/wallet/mod.rs [hunk 17]
```diff
@@ -519,13 +532,76 @@ impl WalletService {
...
+    pub fn import(&self,
+                  pool_name: &str,
+                  name: &str,
+                  storage_type: Option<&str>,
+                  storage_config: Option<&str>,
+                  credentials: &str,
+                  import_config_json: &str) -> Result<(), WalletError> {
+        // TODO - this can be refactor to skip the entire wallet_handle ceremony,
+        // but in order to do that a lot of WalletService needs to be refactored
+        self.create_wallet(pool_name, name, storage_type, storage_config, credentials)?;
```
Comment (vimmerru) [inline libindy/src/services/wallet/mod.rs]: We can create wallet significantly later after the most of validation is done.

File: libindy/src/commands/non_secrets.rs [hunk 18]
```diff
@@ -129,7 +129,14 @@ impl NonSecretsCommandExecutor {
...
         self._check_type(type_)?;
 
-        let res = self.wallet_service.add_record(wallet_handle, type_, id, value, tags_json.unwrap_or("{}"))?; //TODO: question
+        let tags = match tags_json {
+            None => HashMap::new(),
+            Some(tags_string) =>  match serde_json::from_str(tags_string) {
+                Ok(tag_names) => tag_names,
+                Err(serde_json_err) => return Err(IndyError::WalletError(WalletError::InputError(format!("Invalid tags input: {}", tags_string))))
+            }
+        };
```
Comment (Artemkaaas) [inline libindy/src/commands/non_secrets.rs]: Do we really need two level match?

Comment (Artemkaaas) [inline libindy/src/commands/non_secrets.rs]: we always return CommonInvalidStructure in case of failed parsing (on commands layer) of input json parameter.

File: libindy/src/commands/non_secrets.rs [hunk 19]
```diff
@@ -161,7 +168,11 @@ impl NonSecretsCommandExecutor {
 
         self._check_type(type_)?;
 
-        let res = self.wallet_service.update_record_tags(wallet_handle, type_, id, tags_json)?;
+        let tags: HashMap<String, String> = match serde_json::from_str(tags_json) {
+            Ok(tag_names) => tag_names,
+            Err(serde_json_err) => return Err(IndyError::WalletError(WalletError::InputError(format!("Invalid tags input: {}", tags_json))))
+        };
+        let res = self.wallet_service.update_record_tags(wallet_handle, type_, id, &tags)?;
```
Comment (Artemkaaas) [inline libindy/src/commands/non_secrets.rs]: match is not required here and several places below
let tags: <HashMap<String, String>> = serde_json::from_str(tags_json)
            .map_err(|err| CommonError::InvalidStructure(format!("Cannot deserialize tags: {:?}", err)))?;

File: libindy/src/errors/wallet.rs [hunk 20]
```diff
@@ -37,6 +37,9 @@ pub enum WalletError {
     ItemNotFound,
     ItemAlreadyExists,
     QueryError(String),
+    StructureError(String),
+    ExportPathExists,
+    ImportPathDoesNotExist,
 }
```
Comment (Artemkaaas) [inline libindy/src/errors/wallet.rs]: Why do we need this errors?

File: libindy/src/services/wallet/encryption.rs [hunk 21]
```diff
@@ -7,51 +7,60 @@ use utils::crypto::chacha20poly1305_ietf::ChaCha20Poly1305IETF;
...
-                TagName::OfPlain(encrypt_as_searchable(
-                    &tag_name.as_bytes()[1..], tag_name_key, tags_hmac_key))
-            } else {
-                TagName::OfEncrypted(encrypt_as_searchable(tag_name.as_bytes(), tag_name_key, tags_hmac_key))
-            })
-        .collect::<Vec<TagName>>();
-
-    Ok(etag_names)
+    encrypted_tag_names
 }
```
Comment (Artemkaaas) [inline libindy/src/services/wallet/encryption.rs]: Why did you revert these functions?

File: libindy/src/services/wallet/encryption.rs [hunk 22]
```diff
@@ -114,22 +129,50 @@ pub(super) fn decrypt_tags(etags: &Option<Vec<Tag>>, tag_name_key: &[u8], tag_va
...
 mod tests {
     use super::*;
 
     #[test]
     fn test_encrypt_decrypt_tags() {
-        let tags = r#"{"tag1":"value1", "tag2":"value2", "~tag3":"value3"}"#;
+        let mut tags = HashMap::new();
+        tags.insert("tag1".to_string(), "value1".to_string());
+        tags.insert("tag2".to_string(), "value2".to_string());
+        tags.insert("~tag3".to_string(), "value3".to_string());
```
Comment (Artemkaaas) [inline libindy/src/services/wallet/encryption.rs]: I suggest using the sequence of initialization string and then parsing to HashMap instead of building HashMap because it is much more readable and similar to actual usage

File: libindy/src/services/wallet/mod.rs [hunk 23]
```diff
@@ -519,13 +532,76 @@ impl WalletService {
...
+    pub fn export_wallet(&self, wallet_handle: i32, export_config_json: &str, version: u32) -> Result<(), WalletError> {
+        match self.wallets.borrow().get(&wallet_handle) {
+            Some(wallet) => {
+                let export_config: WalletExportConfig = serde_json::from_str(export_config_json)?;
+                let export_path = Path::new(&export_config.path);
+                if export_path.exists() {
+                    return Err(WalletError::ExportPathExists);
+                }
+
+                let export_file = File::create(export_path)?;
```
Comment (Artemkaaas) [inline libindy/src/services/wallet/mod.rs]: use https://doc.rust-lang.org/std/fs/struct.OpenOptions.html instead of manual checking

File: libindy/src/services/wallet/mod.rs [hunk 24]
```diff
@@ -519,13 +532,76 @@ impl WalletService {
...
+                  storage_type: Option<&str>,
+                  storage_config: Option<&str>,
+                  credentials: &str,
+                  import_config_json: &str) -> Result<(), WalletError> {
+        let import_config: WalletExportConfig = serde_json::from_str(import_config_json)?;
+        let import_path = Path::new(&import_config.path);
+        if !import_path.exists() {
+            return Err(WalletError::ImportPathDoesNotExist);
+        }
+
```
Comment (Artemkaaas) [inline libindy/src/services/wallet/mod.rs]: same: https://doc.rust-lang.org/std/fs/struct.OpenOptions.html

File: libindy/src/services/wallet/mod.rs [hunk 25]
```diff
@@ -1434,26 +1561,125 @@ mod tests {
...
-        let expected_tags_json = r#"{"tag_name_2":"new_tag_value_2"}"#;
+        let mut expected_tags = HashMap::new();
+        expected_tags.insert(tag_name_2.to_string(), tag_value_2.to_string());
         let retrieved_tags = item.tags.unwrap();
-        assert_eq!(serde_json::from_str::<Tags>(&expected_tags_json).unwrap(), retrieved_tags);
-    }
+        assert_eq!(expected_tags, retrieved_tags);
+    }
+
+
```
Comment (Artemkaaas) [inline libindy/src/services/wallet/mod.rs]: Why did you add tests that are commented?

File: libindy/src/utils/crypto/chacha20poly1305_ietf/sodium.rs [hunk 26]
```diff
@@ -22,6 +23,14 @@ impl ChaCha20Poly1305IETF {
...
     }
 
+    pub fn increment_nonce(nonce: &mut [u8]) {
+        increment_le(nonce);
+    }
+
+    pub fn nonce_len() -> usize { chacha20poly1305_ietf::NONCEBYTES }
+
+    pub fn key_len() -> usize { chacha20poly1305_ietf::KEYBYTES }
+
```
Comment (Artemkaaas) [inline libindy/src/utils/crypto/chacha20poly1305_ietf/sodium.rs]: Do we need these function? I can find their usages

File: libindy/tests/wallet.rs [hunk 27]
```diff
@@ -231,6 +232,85 @@ mod high_cases {
...
+                fs::create_dir_all(export_dir).unwrap();
+            }
+
+            EnvironmentUtils::tmp_file_path("export_file")
+        }
+
+        fn _prepare_export_wallet_config(path: &str) -> String {
+            format!(r##"{{"key": "{}", "path": "{}"}}"##, "test_key", path)
+        }
+
```
Comment (Artemkaaas) [inline libindy/tests/wallet.rs]: why do you copy/past these functions 4 times?

File: libindy/tests/wallet.rs [hunk 28]
```diff
@@ -231,6 +232,85 @@ mod high_cases {
...
+        #[test]
+        fn indy_export_wallet_works() {
+            TestUtils::cleanup_storage();
+            let export_key = "test_key";
```
Comment (Artemkaaas) [inline libindy/tests/wallet.rs]: this variable doesn't use

File: libindy/tests/wallet.rs [hunk 29]
```diff
@@ -231,6 +232,85 @@ mod high_cases {
...
+            let config_json = _prepare_export_wallet_config(path_str);
+
+            let wallet_name = "indy_export_wallet_works";
+            WalletUtils::create_wallet(POOL, wallet_name, None, None, None).unwrap();
+            let wallet_handle = WalletUtils::open_wallet(wallet_name, None, None).unwrap();
+            WalletUtils::export_wallet(wallet_handle, &config_json).unwrap();
+
+            assert!(path.exists());
+            TestUtils::cleanup_storage();
+        }
```
Comment (Artemkaaas) [inline libindy/tests/wallet.rs]: every test must close wallet after them

File: libindy/tests/wallet.rs [hunk 30]
```diff
@@ -231,6 +232,85 @@ mod high_cases {
...
+            if !export_dir.exists() {
+                fs::create_dir_all(export_dir).unwrap();
+            }
+
+            EnvironmentUtils::tmp_file_path("export_file")
+        }
+
+        fn _prepare_export_wallet_config(path: &str) -> String {
+            format!(r##"{{"key": "{}", "path": "{}"}}"##, "test_key", path)
+        }
```
Comment (Artemkaaas) [inline libindy/tests/wallet.rs]: these have already been declared above.

File: libindy/tests/wallet.rs [hunk 31]
```diff
@@ -231,6 +232,85 @@ mod high_cases {
...
+
+        #[test]
+        fn indy_import_wallet_works() {
```
Comment (Artemkaaas) [inline libindy/tests/wallet.rs]: we need the test that export/import not empty wallet and check compare contents

File: libindy/tests/wallet.rs [hunk 32]
```diff
@@ -460,4 +540,132 @@ mod medium_cases {
...
+
+        #[test]
+        fn indy_export_wallet_returns_error_if_path_exists() {
+            TestUtils::cleanup_storage();
+            let export_key = "test_key";
+            let path = _prepare_export_wallet_path();
+            let path_str = path.to_str().unwrap();
+            fs::File::create(&path).unwrap();
+            let config_json = _prepare_export_wallet_config(path_str);
+
```
Comment (Artemkaaas) [inline libindy/tests/wallet.rs]: Why we can't use helpers functions?

File: libindy/tests/wallet.rs [hunk 33]
```diff
@@ -460,4 +540,132 @@ mod medium_cases {
...
+            let path_str = path.to_str().unwrap();
+            fs::File::create(&path).unwrap();
+
+            let wallet_name = "indy_export_wallet_returns_error_if_invalid_config";
+            WalletUtils::create_wallet(POOL, wallet_name, None, None, None).unwrap();
+            let wallet_handle = WalletUtils::open_wallet(wallet_name, None, None).unwrap();
+            let res = WalletUtils::export_wallet(wallet_handle, "{}");
+
+            // TODO - maybe introduce WalletConfigurationError
+            assert_match!(Err(ErrorCode::WalletDecodingError), res);
```
Comment (Artemkaaas) [inline libindy/tests/wallet.rs]: here must be CommonInvalidStructure

File: libindy/tests/wallet.rs [hunk 34]
```diff
@@ -460,4 +540,132 @@ mod medium_cases {
...
+        fn indy_import_wallet_returns_error_if_path_doesnt_exist() {
+            TestUtils::cleanup_storage();
+            let export_key = "test_key";
+            let path = _prepare_export_wallet_path();
+            let path_str = path.to_str().unwrap();
+            let config_json = _prepare_export_wallet_config(path_str);
+
+            let wallet_name = "indy_import_wallet_returns_error_if_path_doesnt_exist";
+            let res = WalletUtils::import_wallet(POOL, wallet_name, None, None, None, &config_json);
+            assert_match!(Err(ErrorCode::CommonIOError), res);
```
Comment (Artemkaaas) [inline libindy/tests/wallet.rs]: please use assert_eq!(ErrorCode::CommonIOError, res.unwrap_err()); as other tests

File: libindy/src/commands/non_secrets.rs (hunk 18 above)
Comment (keichiri) [reply libindy/src/commands/non_secrets.rs]: you are correct. Your choice is more elegant. I switched to it (and in other places). Thank you

File: libindy/src/services/wallet/encryption.rs (hunk 21 above)
Comment (keichiri) [reply libindy/src/services/wallet/encryption.rs]: Sorry, I didnt look at the content during the merge resolution. It slipped to me that you refactored them meanwhile. I thought it was outdated. I redid the merge, and your changes are active

File: libindy/src/services/wallet/mod.rs [hunk 35]
```diff
@@ -1576,4 +1686,169 @@ mod tests {
         assert_eq!("type", record.get_type().unwrap());
         assert_eq!("value1", record.get_value().unwrap());
     }
-}
+
+    /**
+        Export/Import tests
+    */
+    fn _get_export_dir_path() -> &'static str {
+         "/tmp/indy_wallet_export_tests"
```
Comment (vimmerru) [inline libindy/src/services/wallet/mod.rs]: Will it work on windows and OSX? There is corresponded You can use EnvironmentUtils::tmp_path

File: libindy/src/services/wallet/mod.rs [hunk 36]
```diff
@@ -1576,4 +1686,169 @@ mod tests {
...
+    fn _prepare_export_path() {
+        let export_directory_path = Path::new(_get_export_dir_path());
+        if export_directory_path.exists() {
+            std::fs::remove_dir_all(export_directory_path).unwrap();
+        }
+        std::fs::create_dir(export_directory_path).unwrap();
+    }
+
+    fn _remove_export_path() {
+        let export_directory_path = Path::new(_get_export_dir_path());
```
Comment (vimmerru) [inline libindy/src/services/wallet/mod.rs]: Will TestUtils:: cleanup_storage or cleanup_temp work instead?

File: libindy/src/services/wallet/export_import.rs [hunk 37]
```diff
@@ -0,0 +1,1108 @@
...
+    use serde_json;
+
+    use services::wallet::storage::WalletStorageType;
+    use services::wallet::storage::default::SQLiteStorageType;
+    use services::wallet::wallet::{Keys, Wallet};
+    use services::wallet::encryption::{decrypt_merged, decrypt, encrypt_as_not_searchable};
+    use super::*;
+
+
+    fn _wallet_path() -> std::path::PathBuf {
```
Comment (vimmerru) [inline libindy/src/services/wallet/export_import.rs]: Why don't use EnvironmentUitls ??? There is corresponded helper.

File: libindy/src/services/wallet/export_import.rs [hunk 38]
```diff
@@ -0,0 +1,1108 @@
...
+        let mut path = env::home_dir().unwrap();
+        path.push(".indy_client");
+        path.push("wallet");
+        path.push("test_wallet");
+        path
+    }
+
+    fn _cleanup() {
```
Comment (vimmerru) [inline libindy/src/services/wallet/export_import.rs]: Why don't use TestUtils ???

File: libindy/src/services/wallet/export_import.rs [hunk 39]
```diff
@@ -0,0 +1,1108 @@
...
+        return [
+            1, 2, 3, 4, 5, 6, 7, 8,
+            1, 2, 3, 4, 5, 6, 7, 8,
+            1, 2, 3, 4, 5, 6, 7, 8,
+            1, 2, 3, 4, 5, 6, 7, 8
+        ];
+    }
+
+    fn _create_export_file() -> Box<io::Write> {
+        let path_str = "/tmp/wallet/export_test";
```
Comment (vimmerru) [inline libindy/src/services/wallet/export_import.rs]: Wll it work on windows and OSX? I suggest to use EnvrirnonmentUtils instead.

**Comment by keichiri** [comment]
@vimmerru  I will fix the path issues. 
I am not sure what you mean by commented tests.
There is only one that I see, and that one is very old, not added now
